### PR TITLE
ignore STRICT_DNS/STATIC Cluster doesnot have endpoints

### PR DESCRIPTION
**Pull request by ydh926**

### CODE REVIEW

**Comment by ydh926** [comment]
STRICT_DNS/STATIC Cluster does not have endpoint will block envoy. 
https://github.com/envoyproxy/envoy/issues/9963

Impact on pilot：
https://github.com/istio/istio/issues/20933

**Comment by ydh926** [comment]
@googlebot I fixed it

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 1]
```diff
@@ -201,6 +201,9 @@ func (configgen *ConfigGeneratorImpl) buildOutboundClusters(proxy *model.Proxy,
 			clusterName := model.BuildSubsetKey(model.TrafficDirectionOutbound, "", service.Hostname, port.Port)
 			serviceAccounts := push.ServiceAccounts[service.Hostname][port.Port]
 			defaultCluster := buildDefaultCluster(push, clusterName, discoveryType, lbEndpoints, model.TrafficDirectionOutbound, proxy, port, service.MeshExternal)
+			if defaultCluster == nil {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: how can it happen?

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 2]
```diff
@@ -249,6 +252,9 @@ func (configgen *ConfigGeneratorImpl) buildOutboundClusters(proxy *model.Proxy,
 					lbEndpoints = buildLocalityLbEndpoints(push, networkView, service, port.Port, []labels.Instance{subset.Labels})
 				}
 				subsetCluster := buildDefaultCluster(push, subsetClusterName, discoveryType, lbEndpoints, model.TrafficDirectionOutbound, proxy, nil, service.MeshExternal)
+				if subsetCluster == nil {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: same here

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 3]
```diff
@@ -1303,6 +1312,10 @@ func buildDefaultCluster(push *model.PushContext, name string, discoveryType api
 	}
 
 	if discoveryType == apiv2.Cluster_STATIC || discoveryType == apiv2.Cluster_STRICT_DNS {
+		if direction == model.TrafficDirectionOutbound && len(localityLbEndpoints) == 0 {
+			log.Errorf("STRICT_DNS/STATIC Cluster:%s, does not have endpoints, skipped", name)
+			return nil
+		}
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: Can you explain how can this happen?

Comment (ydh926) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: For example:
1. configure a DNS type ServiceEntry
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: static-20191230
  namespace: gateway-system
spec:
  endpoints:
  - address: 10.108.251.133
    ports:
      http: 80
  - address: 10.96.144.80
    ports:
      http: 80
  exportTo:
  - '*'
  hosts:
  - com.netease.static-20191230
  location: MESH_EXTERNAL
  ports:
  - name: http
    number: 80
    protocol: HTTP
  resolution: DNS
```
2. configure a subset, which cannot select any endpoint
```go
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: static-1230
  namespace: gateway-system
spec:
  host: com.netease.static-20191230
  subsets:
  - altStatName: healthCheckTest
    name: static-20191230-test1
    labels:
       app: none
```

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: IC, seems you should also provide the VirtualService

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 4]
```diff
@@ -1303,6 +1312,10 @@ func buildDefaultCluster(push *model.PushContext, name string, discoveryType api
 	}
 
 	if discoveryType == apiv2.Cluster_STATIC || discoveryType == apiv2.Cluster_STRICT_DNS {
+		if direction == model.TrafficDirectionOutbound && len(localityLbEndpoints) == 0 {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: no need check outbound

Comment (ydh926) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: Done

**Comment by hzxuzhonghu** [comment]
@ydh926 Add a test for it in `cluster_test.go`

**Comment by ydh926** [comment]
> @ydh926 Add a test for it in `cluster_test.go`
is this ok? @hzxuzhonghu

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 5]
```diff
@@ -1303,6 +1312,10 @@ func buildDefaultCluster(push *model.PushContext, name string, discoveryType api
 	}
 
 	if discoveryType == apiv2.Cluster_STATIC || discoveryType == apiv2.Cluster_STRICT_DNS {
+		if len(localityLbEndpoints) == 0 {
+			log.Errorf("STRICT_DNS/STATIC Cluster:%s, does not have endpoints, skipped", name)
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: Remove the error message please as this is going to get very noisy. Instead add it as a metric (look around pilot code for how we do metrics for these kinds of things).

Comment (ydh926) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: Done

File: pilot/pkg/model/push_context.go [hunk 6]
```diff
@@ -429,6 +429,13 @@ var (
 		"Duplicate envoy clusters caused by service entries with same hostname",
 	)
 
+	// StaticNoEndpointClusters tracks dns/static clusters without endpoints
+	StaticNoEndpointClusters = monitoring.NewGauge(
+		"pilot_static_without_endpoints_envoy_clusters",
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/push_context.go]: pilot_dns_cluster_without_endpoints

**Comment by ydh926** [comment]
> One question: we do have a `BlackHoleCluster` cluster of static type without endpoints, why there is no error?

The field loadassignment of BlackHoleCluster is nil，envoy will give it a default value. But it in STRICT_DNS cluster is an array of length 0, envoy will try to resolve it. That is a bug in envoy, but I am not sure when they will fix it.

File: pilot/pkg/model/push_context.go (hunk 6 above)
Comment (ydh926) [reply pilot/pkg/model/push_context.go]: Done

File: pilot/pkg/model/push_context.go [hunk 7]
```diff
@@ -429,6 +429,13 @@ var (
 		"Duplicate envoy clusters caused by service entries with same hostname",
 	)
 
+	// StaticNoEndpointClusters tracks dns/static clusters without endpoints
+	StaticNoEndpointClusters = monitoring.NewGauge(
+		"pilot_dns_cluster_without_endpoints_envoy_clusters",
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/push_context.go]: pilot_dns_cluster_without_endpoints, actually we donot sum up ths static cluster.

**Comment by ramaraochavali** [comment]
I think we should exclude `STATIC` from this check because imagine a case we create a service entry with `STATIC` resolution type and if its endpoints ever become zero, with this change we do not send that cluster to Envoy and Envoy would delete the cluster - That is not correct because we should just clear the endpoints not the cluster it self. I fixed similar thing to allow empty endpoints to static cluster in service entries https://github.com/istio/istio/pull/19036

**Comment by ydh926** [comment]
> I think we should exclude `STATIC` from this check because imagine a case we create a service entry with `STATIC` resolution type and if its endpoints ever become zero, with this change we do not send that cluster to Envoy and Envoy would delete the cluster - That is not correct because we should just clear the endpoints not the cluster it self. I fixed similar thing to allow empty endpoints to static cluster in service entries #19036

In your description, a ServiceEntry of static type will be converted to a cluster of type eds. If so, these changes will not affect the ServiceEntry of static type.

> For Service Entries with STATIC resolution, we should allow empty endpoints. Since we internally convert it to EDS type of cluster in Envoy - it will work and it is valid for the external service represented by the Service Entry to have zero endpoints.

**Comment by hzxuzhonghu** [comment]
make sense, fix the lint, and i will approve.

**Comment by ramaraochavali** [comment]
Ah.. sorry. Makes sense. Thank you

File: pilot/pkg/model/push_context.go (hunk 7 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go]: why this is back

**Comment by ydh926** [comment]
fixed @hzxuzhonghu

**Comment by hzxuzhonghu** [comment]
@ydh926 Do you forget to push?

File: pilot/pkg/model/push_context.go (hunk 7 above)
Comment (ydh926) [reply pilot/pkg/model/push_context.go]: sorry，it is a mistake. fixed already

**Comment by lambdai** [comment]
cherrypick/release-1.4

**Comment by lambdai** [comment]
cherrypick/release-1.5

**Comment by chaudyg** [comment]
@lambdai  I don't think the cherrypick in 1.4 worked. Any idea?

**Pull request merged**